docs/src/proposals/abi-v2.md

```diff
@@ -0,0 +1,88 @@
+
+## Existing Encoding Standards
+Information about each entry (attribute value pair) which is either encoded explicitly or can be implicitly gained from the context:
+- Attribute (semantics) as a tag or name string
+- Data type / encoding
+- Value length
+- Value content
+
+Some [binary data-serialization formats](https://en.wikipedia.org/wiki/Comparison_of_data-serialization_formats#Comparison_of_binary_formats) are [CBOR](https://cbor.io), [Ion](https://amzn.github.io/ion-docs/), [MsgPack](https://msgpack.org), [Protobuf](https://github.com/protocolbuffers/protobuf).
+We currently only have CBOR in our Cargo.lock, so everything else would add a dependency.
+However, none of these support indirections / references / offsets / pointers and they are all in row-major order (see next point). Therefore, we should design and implement our own extensibile serialization format.
+
+## Nesting order
+Borrowing the terms from table based database systems:
+- Row-major encoding (inner loop over attributes, outer loop over accounts) means first the values of all attributes of one account are encoded, then comes the next account.
+- Column-major encoding (inner loop over accounts, outer loop over attributes) means first the values of one attribute for all accounts are encoded, then comes the next attribute.
+
+Column-major encoding has the following advantages in our use case:
+- Extensibility on an attribute level (but not per account level as that would be wasteful).
+- Read / write access control per attribute (not per account meta data, but the account data is handled separately).
+- Splitting meta data into a local instruction context and a global transaction context.
+
+## Design
+This section describes how the new encoding works in general and how we would use it for the new ABI specifically.
+
+### Encoding
+Everything is in little endian.
+
+```Rust
+struct Map {
+  number_of_entries: u16,
+  attribute: [u16; number_of_entries],
+  value_offset: [u32; number_of_entries + 1],
+  is_value_indirect: [u8; number_of_entries],
+  values: [Value],
+}
+
+enum Value {
+  Direct([u8]),
+  Indirect({ pointer: u64, length: u32 }),
+}
+```
+
+### Memory Mapping / Regions
+Additional to the mandatory RBPF memory regions (null, program, stack, heap) we would add `2 + number_of_accounts_in_instruction` memory regions to the VM mapping.
+
+#### Read-only meta data region:
+- Transaction context:
+  - `NumberOfAccountsInTransaction = 1`: `u32`
+  - `AccountKey = 2`: `[Pubkey; NumberOfAccountsInTransaction]`
+  - `AccountIsExecutable = 3`: `[bool; NumberOfAccountsInTransaction]`
+  - `AccountOwner = 4`: `[Pubkey; NumberOfAccountsInTransaction]`
+  - `AccountLamports = 5`: `[u64; NumberOfAccountsInTransaction]`
+  - `AccountData = 6`: `[&[u8]; NumberOfAccountsInTransaction]`
+  - `InvocationStackFrame = 7`: `Map`
+- Instruction context:
+  - `ParentStackFrame = 8`: `Map`
```

Comment:
What is this?

Reply:
Currently the stack is modeled as a linked list (because every stack frame can have a different size). But I will add support for arrays of pointers in values, which is needed for account data anyway. Then we can also refer to an invocation stack frame by index.

Reply:
The invocation stack is now an array of fatpointers.